[20.09] AnnData Peek Fix

mtekman opened the pull request:
This should fix the problematic shape parameter detection for AnnData datasets

Once this is merged, this PR (https://github.com/galaxyproject/tools-iuc/pull/3237) should go green.

mtekman committed:
fix for shape parameter

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -981,12 +981,15 @@ def set_meta(self, dataset, overwrite=True, **kwd):
         try:
             with h5py.File(dataset.file_name, 'r') as anndata_file:
+                ## NOTE: It appears that 'dataset.metadata' variables can
+                ##       only be set exactly once, so tread carefully.
```
```suggestion
```
That does not seem right.

mtekman replied:
Hmm, so maybe my conclusion  was wrong but I am definitely seeing weird behaviour if add 

`dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))`  on L995

The error message I get is:


```
Traceback (most recent call last):
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1547, in _finish_dataset
    dataset.set_peek(line_count=line_count)
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/model/__init__.py", line 2779, in set_peek
    return self.datatype.set_peek(self, **kwd)
TypeError: set_peek() got an unexpected keyword argument 'line_count'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/runners/__init__.py", line 540, in _finish_or_resubmit_job
    job_wrapper.finish(tool_stdout, tool_stderr, exit_code, check_output_detected_state=check_output_detected_state, job_stdout=job_stdout, job_stderr=job_stderr)
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1681, in finish
    self._finish_dataset(
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1550, in _finish_dataset
    dataset.set_peek()
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/model/__init__.py", line 2779, in set_peek
    return self.datatype.set_peek(self, **kwd)
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/datatypes/binary.py", line 1077, in set_peek
    peekstr = "[n_obs x n_vars]\n    %d x %d" % tuple(tmp.shape)
TypeError: 'NoneType' object is not iterable
FAIL
```

So I read this as:

 * `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))`  evaluates to `None` 
 * The `if 'X' in dataset.metadata.layers_names:` block lower down is either never reached, or the `dataset.metadata.shape` variable is never overwritten.

Some more tests:

* Evaluating `util.unicodify(anndata_file.attrs.get('shape'))` does not produce any errors, so it's the assignment to `dataset.metadata.shape` that seems to  cause the problem
* Surrounding `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))` in a try-catch block does not fix the issue, so I think the if-else block at the bottom *is* being reached.

mvdbeek replied:
Can you remove the try/except catching on line 984 / 1058 ?
From what you describe I think the most likely explanation is that you're raising an exception, and the set_peek code relies on no exception having occurred.

That said I don't see `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))` on line 995, has something gone wrong there ? ` File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/datatypes/binary.py", line 1077, in set_peek` also doesn't match the line numbers here.

mtekman replied:
oh I meant "if you add {code snippet} at this line, this is what you get"

I can push this as a commit if you like, but I'd rather leave it out of the git history

mvdbeek replied:
We can always rebase this out.

mtekman replied:
> Can you remove the try/except catching on line 984 / 1058 ?

We should keep this though, right? I took this template from one of the other h5 classes in the file (Loom)

> From what you describe I think the most likely explanation is that you're raising an exception, and the set_peek code relies on no exception having occurred.

though I can see now that this is main try-except block is what was hiding the errors

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -981,12 +981,15 @@ def set_meta(self, dataset, overwrite=True, **kwd):
                 dataset.metadata.doi = util.unicodify(anndata_file.attrs.get('doi'))
                 dataset.creation_date = util.unicodify(anndata_file.attrs.get('creation_date'))
                 # none of the above appear to work in any dataset tested, but could be useful for future
+                # dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))
```
```suggestion
```

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -1040,15 +1047,20 @@ def _layercountsize(tmp, lennames=0):
+                    dataset.metadata.shape = (int(n_obs), int(n_var))
+                else:
+                    log.warning("No shape data could be determined!")
+                    dataset.metadata.shape = tuple(-1,-1)
```
```suggestion
                else:
                    dataset.metadata.shape = (int(dataset.metadata.obs_size or -1), int(dataset.metadata.var_size or -1))
```
I think that should be the core of the changes you're making. You can remove `n_var` and `n_obs` then.
Or you can do
`dataset.metadata.shape = (int(dataset.metadata.obs_size), int(dataset.metadata.var_size))` like you've had it before and set `obs_size ` and `var` size defaults to -1 using the `MetadataElement `definitions.

mvdbeek reviewed (commented):
Can you add the problematic dataset (or a shrunk down variant) in question to lib/galaxy/datatypes/test ?
That will run the upload tests against that file.

mtekman committed:
added anndata test set from tools-iuc/tools/anndata/test-data/

mtekman committed:
By @mvdbeek suggestions:

  The metadata element defaults for the obs and var sizes were set to
  -1, unnecessary variables were removed, further simplifying the
  if-else statements lower in the body since the defaults could now be
  used instead.

mtekman committed:
the default shape parameter was set to (-1, -1)

mtekman committed:
remove try/except block, add default shape get

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -978,14 +980,14 @@ def sniff(self, filename):
                 dataset.metadata.url = util.unicodify(anndata_file.attrs.get('url'))
                 dataset.metadata.doi = util.unicodify(anndata_file.attrs.get('doi'))
                 dataset.creation_date = util.unicodify(anndata_file.attrs.get('creation_date'))
+                dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))
```
`unicodify` in essence turns bytestrings to unicode strings, passing through `None`. `anndata_file.attrs.get('shape')` should be `None` or a tuple, of integers, right ? If you get a tuple from `anndata_file.attrs.get('shape')`  and you pass it into unicodify you'll get a string back out:
```python
In [2]: from galaxy.util import unicodify

In [3]: unicodify((2,2))
Out[3]: '(2, 2)'
```
So that makes me think we have no test yet for a file that specifies the shape.
I would guess that the following is closer to what you want to do:

```suggestion
                dataset.metadata.shape = anndata_file.attrs.get('shape') or dataset.metadata.shape
```
That means you can drop line 1054/1055.

If you decide not to wrap this part in try/except you need to dedent the set_meta function body by 4 spaces.

mtekman replied:
I committed the line and unindented the region, but I kept line 1054/55 since I don't quite see how we're handling cases which have no obvious shape assignment but do have observations and variables.

mtekman committed:
Update lib/galaxy/datatypes/binary.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mtekman committed:
indent fix
